### ISSUE TITLE
i18next falls back to wrong language when using whitelist

### ISSUE THREAD

**Issue by cbikar** [i18next/i18next#738]
Hi,

The issue is similar to #209. We have a whitelist configured that contains the supported languages and have one fallback language ("en"). For most languages we don't distinguish by region, except for Chinese where we use the following setup:

- zh language file -> Contains Chinese simplified
- zh-TW and zh-HK language files -> Contains Chinese traditional

So if the browser language is "zh-TW" i18next should directly load the language file for "zh-TW", but if the browser language is "zh-CN" it should fall back to "zh" and return Chinese simplified.

The problem is that depending on the "load" option, i18next either 

- ignores the region code ("languageOnly") which is not desired as we can't distinguish between Chinese simplified and traditional anymore or
- falls back to the wrong languages ("currentOnly" or "all"). If the browser language is for instance "de-AT" it falls back to "en", because "de-AT" is not in the whitelist (only "de" is).

I think when using the "currentOnly" or "all" setting together with a whitelist, i18next shouldn't immediately use the fallback language if the specific language is not found but rather check if there is only the language (without the region) in the whitelist and then use this one.

Am I missing something in the configuration or is this setup currently not supported?

Thanks!

**Comment by jamuhl** [comment]
why at all you need to define a whitelist? it's not needed at all...

the error you see if i18next loads a file that does not exist is a network error not a javascript error...that's a big difference most people don't get.

**Comment by cbikar** [comment]
Thanks for your prompt response. We try to avoid sending requests that produce a 404 response. Our code is embedded into customers websites and they would immediately ask us to stop sending such requests. That's why we decided to use the whitelist.

**Comment by jamuhl** [comment]
ok i see...i think best options would be to extend the xhr backend or create a custom backend that checks for existence before downloading a file, like:

```
function checkFile(fileUrl) {
    var xmlHttpReq = false;
    var self = this;
    // Mozilla/Safari
    if (window.XMLHttpRequest) {
        self.xmlHttpReq = new XMLHttpRequest();
    }
    // IE
    else if (window.ActiveXObject) {
        self.xmlHttpReq = new ActiveXObject("Microsoft.XMLHTTP");
    }

    self.xmlHttpReq.open('HEAD', fileUrl, true);
    self.xmlHttpReq.onreadystatechange = function() {
        if (self.xmlHttpReq.readyState == 4) {
            if (self.xmlHttpReq.status == 200) {
                alert('the file exists');
            } else if (self.xmlHttpReq.status == 404) {
                alert('the file does not exist');
            }
        }
    }
    self.xmlHttpReq.send();
}

checkFile('/somefile.xml');
```

**Comment by Chris2m** [comment]
I think the point is avoiding a roundtrip to the server when it is not really required. The information about the language files available is already present in the whitelist on the client side.

I would expect that if `de-AT` is not defined in the whitelist i18next would pick `de`as the next best match. Looking at the code this does not seem to require many changes. If you agree conceptually, would you accept a PR to align this flow?

Let me know if I can help.

**Comment by jamuhl** [comment]
@Chris2m not sure...basically this situation is: i want language only but for my region/script i want a special treatment. The cost of a server roundtrip is even on mobile network not high if just checking head.

But anyway...open to some changes allowing your use cases. But i would prefer not breaking current behaviour by adding some additional option `nonExplicitWhitelist` default to false allowing to downgrade de-AT to de....

**Comment by cbikar** [comment]
@jamuhl I created a pull request #741 that introduces the new option and code changes that allow a fallback to language only if the language is not found in the whitelist. Let me know if that's ok or if you had another solution in mind.

**Comment by jamuhl** [comment]
just published i18next@3.3.0
